add test_stackplot in test_datetime.py

QuadroTec opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please provide at least 1-2 sentences describing the pull request in detail
(Why is this change required?  What problem does it solve?) and link to relevant
issues and PRs.
Also please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->

Adds the test_stackplot method in test_datetime.py. The test creates the following stackplots:

1. `datetime` x-axis, stacked numerical y-axis
2. ~~`datetime` x-axis, `datetime` plus stacked `datetime.timedelta` y-axis~~ (not implemented for this pull request)

Stacking multiple `datetime` arrays caused a compile error, because numpy does not support adding two `datetime` objects together. I don't believe it makes much sense to stack `datetime` values in this way, so I have not implemented this test.

This pull request closes the `Axes.stackplot` task in issue [#26859](https://github.com/matplotlib/matplotlib/pull/26859)

**Plot output**
![test_stackplot output](https://github.com/matplotlib/matplotlib/assets/19280099/e0ed0a7e-5404-41ad-902c-a9051c580082)

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->


- [X] Resolves `Axes.stackplot` in [#26864](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

ksunden commented on lib/matplotlib/tests/test_datetime.py:
```diff
@@ -318,11 +318,40 @@ def test_spy(self):
+            [datetime.datetime(2023, 1, i) for i in range(1, N)]
+        )
+        dates_years = np.array(
+            [datetime.datetime(1970 + i, 1, 1) for i in range(1, N)]
```
I would avoid this date, as 1970 can hide problems where we don't actually convert dates properly.

(You are _actually_ avoiding the most likely problematic year by starting at 1, but it is still close enough to throw some alarm bells off)

ksunden commented on lib/matplotlib/tests/test_datetime.py:
```diff
@@ -318,11 +318,40 @@ def test_spy(self):
+
+        ax2.stackplot(
+            dates_days, dates_days, timedelta_days, timedelta_hours, timedelta_seconds
+        )
+        ax2.set_ylim(
+            datetime.datetime(2022, 12, 31),
+            datetime.datetime(2023, 2, 8)
+        )
```
These limits are hiding the fact that the blue section goes all the way to 1970 because of the implicit 0 of `stackplot`.

I would support extending the API of the `baseline` kwarg to accept a scalar value, but that is not how it is implemented now, and I think that is out of the scope of this PR.

As it stands, I think I would rather _not_ test the "timedelta on the y axis" case than provide this test which is not a sensible use of `datetime` (though I think it is not _far_ off, and you made something that looked _okay_... I just think that the tools are not there to make using datetimes sensible here)

I do think `timedelta` is the right unit for this plot, to be clear... just it needs that extra explicit baseline, which is not currently supported (only zero or the various "wiggle" type options are supported currently)

ksunden reviewed (changes requested):
I think this has also uncovered that `stackplot` doesn't quite work as intended for `unit`ful data (which is good, that is why we want these tests, to be clear)

I'll try to write up an issue later about it, but the thing that is throwing alarm bells to me is this line:

https://github.com/matplotlib/matplotlib/blob/978b402159bd40663bcf74d3beef62b009010083/lib/matplotlib/stackplot.py#L81

The `promote_types` call there actually works for datetimes passed as `object` arrays (which is why this test works at all), but _will_ fail for `np.datetime64` arrays (which we should either make work and add to this test or be clear about it not working....) I think that datetimes are a bit of an edge case here, but the broader point about units and that `promote_types` call could be problematic...

QuadroTec commented:
Thanks for the review. I changed the starting year in the first test to 2020 from 1971. I also removed the second test until the baseline issue is addressed.

Just double-checking for the second test -- was the only problem the inability to set an explicit baseline (or any baseline other than `'zero'` for that matter), or were there other issues with how it was written?

ksunden commented:
I think the inability to set a baseline was the only problem for `datetime.datetime` object arrays, but there is a lurking problem for `np.datetime64` arrays, which are usually interoperable with `datetime.datetime`. The test itself was fine, other than the baseline being an implicit 0 mark that does not make sense for the data (and having no way of setting a baseline)

I think that there could be some debate on whether the y axis of this plot should even be a concrete date (rather it is always an offset, perhaps), but we don't actually have handling for that built in yet... 

Again, those shortcomings are of the current implementation, which while being not quite fully broken, is lacking polish, not of the test you wrote.

QuLogic commented on lib/matplotlib/tests/test_datetime.py:
```diff
@@ -318,11 +318,16 @@ def test_spy(self):
-        ax.stackplot(...)
+        mpl.rcParams["date.converter"] = 'concise'
+        N = 10
+        fig, ax = plt.subplots(constrained_layout=True)
```
This should use the new argument:
```suggestion
        fig, ax = plt.subplots(layout='constrained')
```
I'd also move it down near `ax.stackplot` to not split it by the data creation code.

QuLogic commented on lib/matplotlib/tests/test_datetime.py:
```diff
@@ -318,11 +318,16 @@ def test_spy(self):
+        N = 10
+        fig, ax = plt.subplots(constrained_layout=True)
+
+        stacked_nums = np.array([range(1, N)] * 4)
```
Could be written as
```suggestion
        stacked_nums = np.tile(np.arange(1, N), (4, 1))
```

QuadroTec commented:
Hi @QuLogic, I have made the changes you suggested.

QuLogic commented:
Thanks @QuadroTec! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

QuadroTec commented:
Thank you @QuLogic!
